compiler-core/src/error.rs

```diff
@@ -406,54 +416,98 @@ impl Error {
+                             Consider unlocking the responsible locked package(s) :\n{}",
+                            locked_conflicts.iter().map(|s| format!("- {s}")).join("\n")
+                        ),
+                        locked_conflicts,
+                    }
+                } else {
+                    Error::DependencyResolutionFailed(
+                        format!(
+                            "Unable to find compatible versions for the version constraints in your gleam.toml.\n\
+                             The conflicting packages are:\n{}",
+                            conflicting_packages.into_iter().map(|s| format!("- {s}")).join("\n")
+                        )
+                    )
+                }
+            } // end [`ResolutionError::NoSolution`] arm
 
             ResolutionError::ErrorRetrievingDependencies {
                 package,
                 version,
                 source,
-            } => format!(
+            } => {
+                let msg = format!(
                 "An error occurred while trying to retrieve dependencies of {package}@{version}: {source}",
-            ),
+            );
+                Error::DependencyResolutionFailed(msg)
+            }
 
             ResolutionError::DependencyOnTheEmptySet {
                 package,
                 version,
                 dependent,
-            } => format!(
-                "{package}@{version} has an impossible dependency on {dependent}",
-            ),
+            } => {
+                let msg =
+                    format!("{package}@{version} has an impossible dependency on {dependent}",);
+
+                Error::DependencyResolutionFailed(msg)
+            }
 
             ResolutionError::SelfDependency { package, version } => {
-                format!("{package}@{version} somehow depends on itself.")
+                let msg = format!("{package}@{version} somehow depends on itself.");
+                Error::DependencyResolutionFailed(msg)
             }
 
             ResolutionError::ErrorChoosingPackageVersion(err) => {
-                format!("Unable to determine package versions: {err}")
+                let msg = format!("Unable to determine package versions: {err}");
+                Error::DependencyResolutionFailed(msg)
             }
 
             ResolutionError::ErrorInShouldCancel(err) => {
-                format!("Dependency resolution was cancelled. {err}")
+                let msg = format!("Dependency resolution was cancelled. {err}");
+                Error::DependencyResolutionFailed(msg)
             }
 
-            ResolutionError::Failure(err) => format!(
-                "An unrecoverable error happened while solving dependencies: {err}"
-            ),
-        })
+            ResolutionError::Failure(err) => {
+                let default_msg = format!("Dependency resolution was cancelled. {err}");
+                if err.contains(", but it is locked to") {
```

Comment:
I don't think this logic is correct. That error message is created when a locked version falls outside the requested version constraint, but that's not the situation in which we want to offer to unlock packages. We want to unlock if there's a conflict and any of the packages in the conflict are locked to specific versions.

I don't think such logic should be in the error module as it is only concerned with the definition, construction, and displaying of errors. It doesn't know anything about the wider context of the program or why errors would be emitted.